core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala

```diff
@@ -48,32 +55,71 @@ private[spark] class DiskBlockManager(conf: SparkConf, deleteFilesOnStop: Boolea
   // of subDirs(i) is protected by the lock of subDirs(i)
   private val subDirs = Array.fill(localDirs.length)(new Array[File](subDirsPerLocalDir))
 
+  private[spark] val badDirs = ArrayBuffer[File]()
+  private[spark] val dirToBlacklistExpiryTime = new HashMap[File, Long]
+  // Filename hash to dirId, it should be small enough to put into memory
+  private[spark] val migratedDirIdIndex = new ConcurrentHashMap[Int, Int].asScala
+
   private val shutdownHook = addShutdownHook()
 
   /** Looks up a file by hashing it into one of our local subdirectories. */
   // This method should be kept in sync with
   // org.apache.spark.network.shuffle.ExternalShuffleBlockResolver#getFile().
   def getFile(filename: String): File = {
+    var mostRecentFailure: Exception = null
     // Figure out which local directory it hashes to, and which subdirectory in that
     val hash = Utils.nonNegativeHash(filename)
-    val dirId = hash % localDirs.length
+    val dirId = migratedDirIdIndex.getOrElse(hash, hash % localDirs.length)
     val subDirId = (hash / localDirs.length) % subDirsPerLocalDir
 
     // Create the subdirectory if it doesn't already exist
     val subDir = subDirs(dirId).synchronized {
+      // Update blacklist
+      val now = clock.getTimeMillis()
+      val unblacklisted = badDirs.filter(now >= dirToBlacklistExpiryTime(_))
+      unblacklisted.foreach { dir =>
+        badDirs -= dir
+        dirToBlacklistExpiryTime.remove(dir)
+      }
+
       val old = subDirs(dirId)(subDirId)
       if (old != null) {
         old
       } else {
-        val newDir = new File(localDirs(dirId), "%02x".format(subDirId))
-        if (!newDir.exists() && !newDir.mkdir()) {
-          throw new IOException(s"Failed to create local dir in $newDir.")
+        assert(!migratedDirIdIndex.contains(dirId))
```

Comment:
why does `migratedDirIdIndex` use the raw hash (0 to Int.MAX) instead of using the index (0 to localDirs.length)?
This check is against the index, so its inconsistent w/ the rest.  But index seems to make more sense to me in general.  Most importantly, the size will stay bounded -- otherwise it seems that map could go grow quite large, and never gets cleaned up.  (though it does mean you won't evenly redistribute the load)

Reply:
ok, I think I see now -- the problem is that after you expire a disk from the blacklist, you need to know whether a block was relocated or not.

Reply:
@squito Yes, that's it. And I think it should be unlikely that it might grow very large. suppose we would wirte 10000 files, then the size of it might be 10000 * (int type entry size, ~ 10+ bytes), I think it's small enough.
Thanks for careful review.